File: code/Python/75-Sort Colors.py
# -*- coding: utf-8 -*
"""
@author: Kaka Chen
@license: (C) Copyright Personal Uncommercial Limited.
@contact: kakachen1990@gmail.com
@file: 75-Sort Colors.py
@time: 2020/10/7 12:43 下午
@desc:
Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.

Here, we will use the integers 0, 1, and 2 to represent the color red, white, and blue respectively.

Follow up:

Could you solve this problem without using the library's sort function?
Could you come up with a one-pass algorithm using only O(1) constant space?


Example 1:

Input: nums = [2,0,2,1,1,0]
Output: [0,0,1,1,2,2]
Example 2:

Input: nums = [2,0,1]
Output: [0,1,2]
Example 3:

Input: nums = [0]
Output: [0]
Example 4:

Input: nums = [1]
Output: [1]


Constraints:

n == nums.length
1 <= n <= 300
nums[i] is 0, 1, or 2.
"""


class Solution(object):
    # def sortColors(self, nums):
    #     """
    #     :type nums: List[int]
    #     :rtype: None Do not return anything, modify nums in-place instead.
    #     """
    #     n = len(nums)
    #     ptr = 0
    #     for i in range(n):
    #         if nums[i] == 0:
    #             nums[i], nums[ptr] = nums[ptr], nums[i]
    #             ptr += 1
    #     for i in range(ptr, n):
    #         if nums[i] == 1:
    #             nums[i], nums[ptr] = nums[ptr], nums[i]
    #             ptr += 1

    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        if n < 2:
            return
        i, head, tail = 0, 0, n - 1
        while i <= tail:
            if nums[i] == 0:
                nums[i], nums[head] = nums[head], nums[i]
                head += 1
            elif nums[i] == 2:
                nums[i], nums[tail] = nums[tail], nums[i]
                tail -= 1
                i -= 1
            i += 1
        return


if __name__ == '__main__':
    s = Solution()
    ipt = [2,0,2,1,1,0]
    s.sortColors(ipt)
    print(ipt)

File: code/75-Sort Colors/75-Sort Colors.c
/* Sort Colors
Given an array with n objects colored red, white or blue, sort them so that objects of the same color are adjacent, with
the colors in the order red, white and blue.

Here, we will use the integers 0, 1, and 2 to represent the color red, white, and blue respectively.

Note:
You are not suppose to use the library's sort function for this problem.

click to show follow up.

Follow up:
A rather straight forward solution is a two-pass algorithm using counting sort.
First, iterate the array counting number of 0's, 1's, and 2's, then overwrite array with total number of 0's, then 1's
and followed by 2's.

Could you come up with an one-pass algorithm using only constant space?
*/


void swap(int * a, int * b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

void sortColors(int* nums, int numsSize){
    int red = 0, blue = numsSize - 1;
    for (int i = 0; i <= blue;) {
        if (nums[i] == 0) {
            swap(&nums[i ++], &nums[red ++]);
        } else if (nums[i] == 2) {
            swap(&nums[i], &nums[blue --]);
        } else {
            i ++;
        }
    }
}